buffer.go

```diff
@@ -0,0 +1,243 @@
+	} else if _, exist := c.RenameTo[eventId]; exist {
+		return true
+	}
+	return false
+}
+func (c *Cache) Add(e Event, mode string) {
+	eventId := e.ID
+	if mode == "RENAME_TO" {
+		c.cacheLock.RLock()
+		renameFromEvent, exist := c.RenameFrom[eventId-1]
+		c.cacheLock.RUnlock()
+
+		if exist {
+			c.createRenameEvent(renameFromEvent, e, exist, true)
+			c.cacheLock.Lock()
+			delete(c.RenameFrom, renameFromEvent.ID)
+			c.cacheLock.Unlock()
+		} else {
+			c.cacheLock.Lock()
+			c.RenameTo[e.ID] = e
+			c.addToTimer(e.ID)
+			c.cacheLock.Unlock()
+		}
+	} else {
+		c.cacheLock.RLock()
+		renameToEvent, exist := c.RenameTo[eventId+1]
+		c.cacheLock.RUnlock()
+
+		if exist {
+			c.createRenameEvent(e, renameToEvent, true, exist)
+			c.cacheLock.Lock()
+			delete(c.RenameTo, renameToEvent.ID)
+			c.cacheLock.Unlock()
+		} else {
+			c.cacheLock.Lock()
+			c.RenameFrom[e.ID] = e
+			c.addToTimer(e.ID)
+			c.cacheLock.Unlock()
+		}
+	}
+}
+func (c *Cache) getEvent(eventId uint64, mode string) (reqEvent Event, exist bool) {
+
+	if mode == "RENAME_TO" {
+		if reqEvent, exist = c.RenameTo[eventId]; exist {
+			c.cacheLock.Lock()
+			delete(c.RenameTo, reqEvent.ID)
+			c.cacheLock.Unlock()
+		}
+		return
+	} else {
+		if reqEvent, exist = c.RenameFrom[eventId]; exist {
+			c.cacheLock.Lock()
+			delete(c.RenameFrom, reqEvent.ID)
+			c.cacheLock.Unlock()
+		}
+		return
+	}
+}
+
+/*
+Check for expired events
+3 types of event are in the cache
+1- the ones that have been matched
+2- the unmatched ones with expired timer
+3- the unmatched ones waiting for its match
+This funtion keep removing the first event in the list untill it reaches the 3rd case
+*/
+func (c *Cache) findExpiredEvents() {
+	itemRemoved := true
+	for c.timers.NumberOfItems > 0 && itemRemoved {
+		if c.timeDifference(c.timers.Head().T)*1000 > int(1*time.Millisecond) {
+			//The item is expired
+			if c.EventExists(c.timers.Head().ID) {
+				event, matchedEvent, eventExist, matchExist, eventType := c.CheckForMatch(c.timers.Head().ID)
+				if eventType == "RENAME_TO" {
+					c.createRenameEvent(matchedEvent, event, matchExist, eventExist)
+				} else {
+					c.createRenameEvent(event, matchedEvent, eventExist, matchExist)
+				}
+			}
+			c.cacheLock.Lock()
+			c.timers.moveHead()
+			c.cacheLock.Unlock()
+		} else if !c.EventExists(c.timers.Head().ID) {
+			//check the first case
+			c.cacheLock.Lock()
+			c.timers.moveHead()
+			c.cacheLock.Unlock()
+		} else {
+			//third case happened
+			itemRemoved = false //break the loop if the head has not changed
+		}
+	}
+}
+
+func (c *Cache) BroadcastRenameEvent(e Event) {
+	events := make([]Event, 1)
+	events[0] = e
+	c.ProcessedEvents <- events
+}
+
+func (c *Cache) Start(es *EventStream) {
+	c.RenameFrom = make(map[uint64]Event)
+	c.RenameTo = make(map[uint64]Event)
+
+	c.timers = TimerCache{make([]TimerCacheItem, MaxCapacity), 0, MaxCapacity, 0, 0}
+	c.ProcessedEvents = es.Events
+
+	ticker := time.NewTicker(100 * time.Millisecond)
```

Comment:
Thinking more I'm a little concerned about this. We're basically polling every 100ms, and we've been bitten by that in the past. @hu13 any thoughts on that?